### ISSUE TITLE
docker login on google container registry

### ISSUE THREAD

**Issue by peterklipfel** [apocas/dockerode#285]
I can not seem to find the right combination of auth variables for the google container registry. The commands on the [google container registry docs](https://cloud.google.com/container-registry/docs/advanced-authentication) all work. But when I put those parameters into an auth object in dockerode, they don't work.

```

console.log(process.env["GCLOUD_JSON_KEY"])
var auth = {
    username: "_json_key",
    password: process.env["GCLOUD_JSON_KEY"],
    email: "not@val.id",
    serveraddress: "https://gcr.io",
    auth: ""
}

```

and

```

var auth = {
    username: "oauth2accesstoken",
    password: "theOauthTokenThatWorksWithDockerLogin",
    email: "not@val.id",
    serveraddress: "https://gcr.io",
    auth: ""
}

```

and I also tried copying the generated base64 encoded key out of my docker config, 

```

var auth = { key: "TheCopiedKey" }

```

The error that I get back is: `{"errorDetail":{"message":"unauthorized: authentication required"},"error":"unauthorized: authentication required"}`

I also tried several variations of mixing and matching. Nothing to this point has worked. Any pointers?

**Comment by Seikho** [comment]
I was able to get this working:
```js
const auth = {
  username: '_json_key',
  password: fs.readFileSync('keyfile.json').toString(),
  serveraddress: 'gcr.io'
}

new Docker(...).checkAuth(auth, (err, res) => {
  if (err) console.log('Error', err);
  console.log(res);
});

// { Status: 'Login Succeeded' }
```

For push or pull, it should be:
```js
// assumes 'const auth = { ...} from before...
const options = { authconfig: auth };
client // client = new Docker(...)...
  .getImage('gcr.io/your-project-id/image:tag')
  .push(options, (err, res) => {
    if (err) return console.log('Error', err);
    res.on('data', d => console.log(d.toString());
});
```

**Comment by apocas** [comment]
Can we close this one? :)

**Comment by peterklipfel** [comment]
Sorry for the delay. Yes. This can be closed. 

The credentials were actually working, but it when I had to push a new image, google doesn't seem to use the cache unless I'm using the same docker session that I was using on the previous push of a similar image. 

So every time I restarted the server, it would push gigabytes of docker. Then, networking throttling would cause the job to fail. If I could get an image to push all the way, then the subsequent pushes of similar images were super fast, so it seemed nondeterministic.

This was coupled with my own config issues (it was late).

My follow on question is whether the caching thing is a gcr/docker registry thing or something that I can fix with my own docker configs.
